tests/test_client.py

```diff
@@ -0,0 +1,258 @@
+
+        ad_token = client.encrypt(example_uid)
+
+        result = client.decrypt(ad_token)
+        self.assertEqual(example_uid, result.uid2)
+
+    @patch('uid2_client.client.post')
+    def test_cannot_encrypt_if_no_key_from_default_keyset(self, mock_post):
+        def get_post_refresh_keys_response_with_no_default_keyset_key(base_url, path, headers, data):
+            response_payload = key_set_to_json_for_sharing([master_key]).encode()
+            return self._make_post_response(data, response_payload)
+
+        mock_post.side_effect = get_post_refresh_keys_response_with_no_default_keyset_key
+        client = Uid2ClientFactory.create("base_url", "api_key", client_secret)
+        client.refresh_keys()
+
+        with self.assertRaises(EncryptionError) as context:
+            client.encrypt(example_uid)
+            self.assertTrue('No Site ID in keys' in context.exception)
+
+    @patch('uid2_client.client.post')
+    def test_cannot_encrypt_if_theres_no_default_keyset_header(self, mock_post):
+        def get_post_refresh_keys_response_with_no_default_keyset_header(base_url, path, headers, data):
+            response_payload = key_set_to_json_for_sharing_with_header("", site_id, [master_key, site_key]).encode()
+            return self._make_post_response(data, response_payload)
+
+        mock_post.side_effect = get_post_refresh_keys_response_with_no_default_keyset_header
+        client = Uid2ClientFactory.create("base_url", "api_key", client_secret)
+        client.refresh_keys()
+
+        with self.assertRaises(EncryptionError) as context:
+            client.encrypt(example_uid)
+            self.assertTrue('No Keyset Key Found' in context.exception)
+
+    @patch('uid2_client.client.post')
+    def test_expiry_in_token_matches_expiry_in_response(self, mock_post):
+        def get_post_refresh_keys_response_with_token_expiry(base_url, path, headers, data):
+            response_payload = key_set_to_json_for_sharing_with_header('"default_keyset_id": 99999, '
+                                                                       '"token_expiry_seconds": 2,', 99999, [master_key,
+                                                                                                             site_key]).encode()
+            return self._make_post_response(data, response_payload)
+
+        mock_post.side_effect = get_post_refresh_keys_response_with_token_expiry
+        client = Uid2ClientFactory.create("base_url", "api_key", client_secret)
+        client.refresh_keys()
+
+        ad_token = client.encrypt(example_uid)
+
+        result = client.decrypt(ad_token)
+        self.assertEqual(example_uid, result.uid2)
+
+        real_decrypt_v3 = encryption._decrypt_token_v3
+
+        with patch('uid2_client.encryption._decrypt_token_v3') as mock_decrypt:
+            def decrypt_side_effect(token_bytes, keys, now):
+                return real_decrypt_v3(token_bytes, keys, now + dt.timedelta(seconds=2))
```

Comment:
when i ran this test, it actually failed. then it passed a few times so not very deterministic. From what i understand, it's trying to verify that the token expiry is enforced. Could we add some buffer here to make sure the test always passes?

Or could you mock the clock/now function used during encryption/decryption so that this test is completely deterministic?

Reply:
increased to 3s and not getting any failures. python does not seem to make it easy to mock functions written in C like now(). Probably not worth implementing for a single test case.